### ISSUE TITLE
[BUG] Warning thrown up when running multiple 21cmFAST models in parallel

### ISSUE THREAD

**Issue by BradGreig** [21cmfast/21cmFAST#38]
**Describe the bug:**
When running multiple 21cmFAST instances I receive the following warning:

2019-10-03 19:07:17,066 | WARNING | _utils.py::write() | pid=18236 | When attempting to write {} to file, write failed with the following error. Continuing without caching.
2019-10-03 19:07:17,067 | WARNING | _utils.py::write() | pid=18236 | Unable to create file (unable to lock file, errno = 35, error message = 'Resource temporarily unavailable')

**To Reproduce:**
See the below MWE

```
import os
import matplotlib.pyplot as plt
import numpy as np
import multiprocessing
import h5py


import logging

from py21cmfast import UserParams, CosmoParams, FlagOptions, AstroParams, run_lightcone, run_coeval

from py21cmfast.c_21cmfast import ffi, lib
from py21cmfast._utils import StructWithDefaults, OutputStruct as _OS, StructInstanceWrapper, StructWrapper

global_params = StructInstanceWrapper(lib.global_params, ffi)

import ctypes
from ctypes import cdll
from ctypes import c_float,c_int,c_uint,c_double,c_char_p

SharedLibraryLocation = "%s/Compute_21cmPS.so"%(os.getcwd())

Lib_21cmPS = cdll.LoadLibrary(SharedLibraryLocation)
Function_21cmPS = Lib_21cmPS.Compute_21cmPS

class Point(ctypes.Structure):
    _fields_ = [('PSbins', c_int),
        ('PS_k', ctypes.POINTER(c_float)),
        ('PS', ctypes.POINTER(c_float)),
        ('PS_error', ctypes.POINTER(c_float))        
        ];

Function_21cmPS.restype = Point
Function_21cmPS.argtypes = [c_int, c_int, c_int, c_float, ctypes.POINTER(c_float)]

user_params = UserParams(DIM=200,HII_DIM=50,BOX_LEN=150.0,USE_FFTW_WISDOM=True,HMF=1)
flag_options = FlagOptions(USE_MASS_DEPENDENT_ZETA=True,USE_TS_FLUCT=True,INHOMO_RECO=True,SUBCELL_RSD=True,M_MIN_in_Mass=True,PHOTON_CONS=True)

def worker(iterable, redshift, random_seed_list, cosmo_params_list, astro_params_list):

    cosmo_params = cosmo_params_list[iterable]
    astro_params = AstroParams(astro_params_list[iterable],INHOMO_RECO=flag_options.INHOMO_RECO)

    LC = run_lightcone(redshift=redshift,
                        max_redshift=global_params.Z_HEAT_MAX,
                        user_params=user_params,
                        cosmo_params=cosmo_params,
                        astro_params=astro_params,
                        flag_options=flag_options,
                        regenerate=True,
                        write=False,
                        z_step_factor=global_params.ZPRIME_STEP_FACTOR,
                        use_interp_perturb_field=True,
                        random_seed=random_seed_list[iterable]
                        )
    
    with h5py.File('data_%s.h5'%(iterable), 'w') as f:

        f.attrs['random_seed'] = random_seed_list[iterable]

        for nm, tp in ffi.typeof(cosmo_params_list[iterable]._cstruct[0]).fields:
            f.attrs['%s'%(nm)] = getattr(cosmo_params_list[iterable],nm)

        for nm, tp in ffi.typeof(astro_params_list[iterable]._cstruct[0]).fields:
            f.attrs['%s'%(nm)] = getattr(astro_params_list[iterable],nm)            

        for nm, tp in ffi.typeof(flag_options._cstruct[0]).fields:
            f.attrs['%s'%(nm)] = getattr(flag_options,nm)            

        for nm, tp in ffi.typeof(user_params._cstruct[0]).fields:
            f.attrs['%s'%(nm)] = getattr(user_params,nm)        

        for nm, tp in ffi.typeof(global_params._cobj).fields:
            if 'external_table_path' in nm:                
                f.attrs['%s'%(nm)] = ffi.string(getattr(global_params,nm))
            else:
                f.attrs['%s'%(nm)] = getattr(global_params,nm)      

        f.create_dataset('redshifts_global', data=LC.node_redshifts, compression="gzip", compression_opts=7, shuffle=True)
        f.create_dataset('neutral_fraction_global', data=LC.global_xHI, compression="gzip", compression_opts=7, shuffle=True)
    
if __name__ == '__main__':

    redshift = 6.0

    num_models = 8

    cosmo_params_list = []
    astro_params_list = []
    random_seed_list = []
    
    for i in range(num_models):

        astro_params_list.append(
            AstroParams(
                ALPHA_ESC=np.random.uniform(-1.,0.5), 
                ALPHA_STAR=np.random.uniform(-0.5,1.), 
                F_ESC10=np.random.uniform(-3.,0.), 
                F_STAR10=np.random.uniform(-3.,0.), 
                L_X=np.random.uniform(38.,42.), 
                M_TURN=np.random.uniform(8.,10.), 
                NU_X_THRESH=np.random.uniform(100.,1500.), 
                X_RAY_SPEC_INDEX=np.random.uniform(-2.,2.),
                t_STAR=np.random.uniform(0.,1.)
                ))
        
             
        random_seed_list.append(42)

        cosmo_params_list.append(
            CosmoParams(            
            OMb=0.0486, 
            OMm=0.3075, 
            POWER_INDEX=0.97, 
            SIGMA_8=0.82, 
            hlittle=0.6774
            ))

       
    num_processes = 8
    num_divisions = int(np.floor(num_models/num_processes))

    for i in range(num_divisions):
        processes = []

        for ii in range(num_processes):
            p = multiprocessing.Process(target=worker, args=(ii + num_processes*i,redshift,random_seed_list, cosmo_params_list, astro_params_list))
            p.start()
            processes.append(p)

        for p in processes:
            p.join()

    remainder = num_models%num_processes

    processes = []

    for ii in range(remainder):
        p = multiprocessing.Process(target=worker, args=(ii + num_divisions*num_processes,6.0,random_seed_list, cosmo_params_list, astro_params_list))
        p.start()
        processes.append(p)

    for p in processes:
            p.join()
```    

**Expected behavior:**

**Details:**
 - Python Version: [e.g. 22]
 - Relevant compiler versions:
 - Relevant package versions:

**Additional context**

**Comment by BradGreig** [comment]
Hey @steven-murray, I can confirm that this error still occurs. 

The issue is that if h5py is writing to file and another processor attempts to write to file (it can be a completely different file, it'll still cause this error) then without a filelock it'll crash.

One possible solution I found was to mimic this [solution](https://github.com/keras-team/keras/issues/11101#issuecomment-555194244). I implemented it into the `write()` function of `_utils.py` and can confirm that it does indeed resolve the issue.

Now the question is, is such a solution acceptable? If the amount of time spent writing to file is sufficiently small, then I think this is fine. However, if the amount of time is large, essentially sequential file writing is occurring.

My script above generated this error as I am running many models simultaneously (and storing to h5 format). Say for database generation or equivalent. Its not an issue for 21CMMC as no data is stored in h5 format. Also, obviously not an issue for single runs either.

**Comment by steven-murray** [comment]
Hey @BradGreig, that kind of makes sense, although it's very strange if HDF5 won't open two _different_ files for writing at the same time...

I think the workaround is perfectly fine, given the fact that this shouldn't happen that much, and that writing shouldn't take too long.  However, I'd be concerned if different files aren't able to be written at the same time. Can you open a PR with your fix in `write()`, and link to this issue? I'll have a quick look around to check if writing to different files should be an issue.

**Comment by steven-murray** [comment]
I just did a quick test using different files using the following script:

```python
import numpy as np
import multiprocessing
import h5py

def writefile(i):
    with h5py.File(f'tempfile{i}.h5', 'w') as fl:
        grp = fl.create_group("group")
        grp['data'] = np.ones(100000000)


p = multiprocessing.Pool(8)
p.map(writefile, range(8))
```

It works perfectly fine. If I take out the `{i}` form the filename (so that it's the same filename in each process), I get the `resource temporarily unavailable` error).

We might get interesting errors though, if we're trying to write to the same file in different processes. Even if we sleep until we can technically write to it, do we really want to do that? I mean, that would mean you just calculated the same data twice...

I guess it's not an error so much as a problem with the way the user set up the runs.

**Comment by BradGreig** [comment]
Hey @steven-murray, I dug deeper into this. The original source of this problem was resolved with #92. Files were being written that should not have been (and those had the same parameters, thus were given the same filenames). 

So nothing needs to be done for this, so I'll mark this as closed.

**Comment by steven-murray** [comment]
Thanks @BradGreig. Wasn't that issue fixed a year ago, though? How were you encountering it now?

**Comment by BradGreig** [comment]
Hey @steven-murray, the original issue that I proposed was solved with #92. I was able to replicate it by setting `write=True`. The specific scenario I was encountering this for was generating various astrophysical models that used the same ICs. Previously, the ICs were being written when they shouldn't have been, thus #92 fixed this. By setting `write=True` it replicated the same problem as all processes were attempting to write the same ICs.

I then realised (forgot...) that that the unique filename generated for the ICs (and perturb field) do not depend on the astrophysical parameters. And thus all were trying to write to the same filename.

So #92 was the fix, and then my recent replication is effectively an enforced user error.
